optuna/samplers/grid.py

```diff
@@ -0,0 +1,183 @@
+        :func:`~optuna.study.Study.optimize` method.
+
+    Note:
+
+        :class:`~optuna.samplers.GridSampler` does not take care of a parameter's quantization
+        specified by discrete suggest methods but just samples one of values specified in the
+        search space. E.g., in the following code snippet, either of ``-0.5`` or ``0.5`` is
+        sampled as ``x`` instead of an integer point.
+
+        .. testcode::
+
+            import optuna
+
+            def objective(trial):
+                # The following suggest method specifies integer points between -5 and 5.
+                x = trial.suggest_discrete_uniform('x', -5, 5, 1)
+                return x ** 2
+
+            # Non-int points are specified in the grid.
+            search_space = {'x': [-0.5, 0.5]}
+            study = optuna.create_study(sampler=optuna.samplers.GridSampler(search_space))
+            study.optimize(objective, n_trials=2)
+
+    Args:
+        search_space:
+            A dictionary whose key and value are a parameter name and the corresponding candidates
+            of values, respectively.
+    """
+
+    def __init__(self, search_space):
+        # type: (Dict[str, List[GridValueType]]) -> None
+
+        for param_name, param_values in search_space.items():
+            for value in param_values:
+                self._check_value(param_name, value)
+
+        self._search_space = collections.OrderedDict()
+        for param_name, param_values in sorted(search_space.items(), key=lambda x: x[0]):
+            self._search_space[param_name] = sorted(param_values)
+
+        self._all_grids = list(itertools.product(*self._search_space.values()))
+        self._param_names = sorted(search_space.keys())
+        self._n_min_trials = len(self._all_grids)
+
+    def infer_relative_search_space(self, study, trial):
+        # type: (Study, FrozenTrial) -> Dict[str, BaseDistribution]
+
+        return {}
+
+    def sample_relative(self, study, trial, search_space):
+        # type: (Study, FrozenTrial, Dict[str, BaseDistribution]) -> Dict[str, Any]
+
+        # Instead of returning param values, GridSampler puts the target grid id as a system attr,
+        # and the values are returned from `sample_independent`. This is because the distribution
+        # object is hard to get at the beginning of trial, while we need the access to the object
+        # to validate the sampled value.
+
+        unvisited_grids = self._get_unvisited_grid_ids(study)
+
+        if len(unvisited_grids) == 0:
+            raise ValueError('All grids have been evaluated. If you want to avoid this error, '
+                             'please make sure that unnecessary trials do not run during '
+                             'optimization by properly setting `n_trials` in `study.optimize`.')
+
+        # In distributed optimization, multiple workers may simultaneously pick up the same grid.
+        # To make the conflict less frequent, the grid is chosen randomly.
```

Comment:
Does this imply that in case of conflicts/race condition, that the whole grid isn't explored? I believe this could become an issue if each trial takes long time. Please correct me if I'm wrong.

Reply:
Maybe we shouldn't raise after finishing the entire grid after all. 🤔

Reply:
A conflict may cause duplicated search for the same point. Even in that case, whole grid would be searched **if enough `n_trials` is passed**. If the number of grid points is `N` and conflict happens `M` times, `n_trials` should be `N` + `M`.

There is another problem, though, that it's hard to know appropriate `n_trials` when specifying `n_trials`, which should be followed up as mentioned in https://github.com/optuna/optuna/pull/665#discussion_r361424127

Reply:
Sorry maybe I wasn't clear enough but what you're describing is sort of what I meant, that you cannot know `N` + `M` beforehand. But I see understand it's a future TODO. Thank you for the explanation.